### PR TITLE
feat: Introduce RBAC based approach to pod logs #7211

### PR DESCRIPTION
**Pull request by reggie-k**
The required changes in the cli were performed, the UI changes will come in a separate PR.
The enforcement is disabled by default, to mitigate the breaking change.
The switch will reside in the main argocd-cm.
Not sure whether I should document the switch explicitly myself or will it be documented in the release blog?

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [V] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [V] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [V] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [V] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [V] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [V] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: server/account/account.go [hunk 1]
```diff
@@ -126,6 +126,18 @@ func (s *Server) CanI(ctx context.Context, r *account.CanIRequest) (*account.Can
 	if !slice.ContainsString(rbacpolicy.Resources, r.Resource, nil) {
 		return nil, status.Errorf(codes.InvalidArgument, "%v does not contain %s", rbacpolicy.Resources, r.Resource)
 	}
+
+	// Temporarily, logs RBAC will be enforced only if an intermediate env var serverRBACLogEnforceEnable is defined and has a "true" value
+	// Otherwise, no RBAC enforcement for logs will take place (meaning, can-i request on a logs resource will result in "yes")
+	// In the future, logs RBAC will be always enforced and the parameter along with this check will be removed
+	if r.Resource == "logs" {
+		// logsRBACEnforceEnable := env.ParseBoolFromEnv("ARGOCD_SERVER_RBAC_LOG_ENFORCE_ENABLE", false)
```
Comment (pasha-codefresh) [inline server/account/account.go:134]: i think it is redundant changes

File: server/application/application.go [hunk 2]
```diff
@@ -1196,6 +1196,16 @@ func (s *Server) PodLogs(q *application.ApplicationPodLogsQuery, ws application.
 		return err
 	}
 
+	// Temporarily, logs RBAC will be enforced only if an intermediate env var serverRBACLogEnforceEnable is defined and has a "true" value
+	// Otherwise, no RBAC enforcement for logs will take place (meaning, can-i request on a logs resource will result in "yes")
+	// In the future, logs RBAC will be always enforced and the parameter along with this check will be removed
+	logsRBACEnforceEnable, _ := s.settingsMgr.GetServerRBACLogEnforceEnable()
```
Comment (pasha-codefresh) [inline server/application/application.go:1202]: not ignore error here

File: util/settings/settings.go [hunk 3]
```diff
@@ -630,6 +634,17 @@ func (mgr *SettingsManager) GetPasswordPattern() (string, error) {
...
 
+func (mgr *SettingsManager) GetServerRBACLogEnforceEnable() (bool, error) {
+	argoCDCM, err := mgr.getConfigMap()
+	if err != nil {
+		return false, err
+	}
+	if argoCDCM.Data[settingsServerRBACLogEnforceEnableKey] == "" {
+		return false, err
+	}
+	return argoCDCM.Data[settingsServerRBACLogEnforceEnableKey] == "true", nil
```
Comment (pasha-codefresh) [inline util/settings/settings.go:645]: maybe better use strconv.ParseBool(

File: util/settings/settings.go [hunk 4]
```diff
@@ -91,6 +91,8 @@ type ArgoCDSettings struct {
 	PasswordPattern string `json:"passwordPattern,omitempty"`
 	// BinaryUrls contains the URLs for downloading argocd binaries
 	BinaryUrls map[string]string `json:"binaryUrls,omitempty"`
+	// ServerRBACLogEnforceEnable temporary var indicates whether rbac will be enforced on logs
+	ServerRBACLogEnforceEnable bool `json:"serverRBACLogEnforceEnable"`
```
Comment (pasha-codefresh) [inline util/settings/settings.go:95]: i dont think you need it anymore after migrate to settings manager

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

**Comment by reggie-k** [comment]
Fixed according to the remarks.
Signed off the commits.
Perhaps additional discussion needed on [util/settings/settings.go](https://github.com/argoproj/argo-cd/pull/8353/files/97222fc22f7495b64436b02c3819c87523f687da#diff-3b21889f7a356a28ebba12c6d339b945d7171a83b4dd603c0902e62cee3051ff)

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Could you also do a quick update in rbac.md? Seems like a good place to document this.

File: server/account/account.go [hunk 5]
```diff
@@ -126,6 +126,21 @@ func (s *Server) CanI(ctx context.Context, r *account.CanIRequest) (*account.Can
 	if !slice.ContainsString(rbacpolicy.Resources, r.Resource, nil) {
 		return nil, status.Errorf(codes.InvalidArgument, "%v does not contain %s", rbacpolicy.Resources, r.Resource)
 	}
+
+	// Temporarily, logs RBAC will be enforced only if an intermediate env var serverRBACLogEnforceEnable is defined and has a "true" value
```
[the remark this answers is not in the document]
Comment (reggie-k) [reply server/account/account.go:130]: Internal var based on a value of an env var. Will clarify the comment, thanks.

[the remark this answers is not in the document]
Comment (reggie-k) [reply server/account/account.go]: Done

**Review by crenshaw-dev — APPROVED** [review]
LGTM!

**Comment by reggie-k** [comment]
Great!
As for the UI changes needed (both backend and UI changes will have to appear in a release), should they come after this PR is merged or can they be based on my branch as it is now?  They will be performed by my colleague.

**Review by pasha-codefresh — APPROVED** [review]

**Comment by pasha-codefresh** [comment]
Well done @reggie-k

**Comment by yonimoses** [comment]
> Great! As for the UI changes needed (both backend and UI changes will have to appear in a release), should they come after this PR is merged or can they be based on my branch as it is now? They will be performed by my colleague.

Awsome!

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Oh! You asked a question that I failed to answer. `settingsServerRBACLogEnforceEnableKey` should be documented in docs/operator-manual/argocd-cm.yaml.

Nitpick: could you replace the hard-coded instances of the config key in tests with the variable?

**Comment by reggie-k** [comment]
> > As for the UI changes needed (both backend and UI changes will have to appear in a release), should they come after this PR is merged or can they be based on my branch as it is now?
> 
> If the backend changes shouldn't be allowed to go out without frontend changes, I'd recommend making them all part of one PR.

I see the reason behind this. How can this be achieved, given my colleague will work on the UI changes, and will need the backend to have my changes? Can you guide me through the process from Github perspective? Where does he fork from, where does he create PRs to, etc?

**Comment by crenshaw-dev** [comment]
> Done, now wondering whether part of that belongs in the release blog :) What do you think?

I think it probably does. @alexmt do we keep a running list anywhere of things that need to be in the release blog?

> Am a total go newbie, but the const key names in settings.go are private and used within the package, as far as I understand, do you want me to make this particular one a global const so that it can be referenced outside?

I'm relatively new to go as well, so I missed that problem. 😆 I think hard-coding is fine in this case. Searching text is easy.

> Can you guide me through the process from Github perspective? Where does he fork from, where does he create PRs to, etc?

I think he could 1) fork argo-cd, 2) add your fork as a remote, 3) checkout your branch, 4) create a new branch based on yours, 5) make changes and then open a PR against your branch. Once the UI changes are merged into your branch, this PR will be ready for final review/merge.

**Comment by alexmt** [comment]
Thank you for review @pasha-codefresh and @crenshaw-dev  . Doing one final round and merging

**Review by alexmt — APPROVED** [review]
LGTM. 


I think we should resolve https://github.com/argoproj/argo-cd/pull/8353/files#r798567363 but I propose to merge PR for the sake of time and create follow up PR to do it.

**Review by crenshaw-dev — APPROVED** [review]
LGTM!

**Comment by pasha-codefresh** [comment]
Great job @reggie-k !

**Comment by alexmt** [comment]
Thank you @reggie-k !

**Pull request merged**